liblxqt-config-cursor/man/lxqt-config-input.1
.TH LXQT-CONFIG-INPUT "1" "January 2025" "LXQt 2.1.0" "LXQt Input Settings"
.SH NAME
lxqt-config-input \- Application of \fBLXQt\fR: The Lightweight Qt Desktop
Environment
.SH SYNOPSIS
.B lxqt-config-input
.br
.SH DESCRIPTION
This is a simple GUI application for keyboard, mouse, and touchpad settings.
(Currently it can only run under X11.)
.P
.SH OPTIONS
-s, --show-page <name>  Choose the page to be shown.
.br
--load-touchpad         Load last touchpad settings.
.br
-v, --version           Displays version information.
.br
-h, --help              Displays help on command line options.
.br
--help-all              Displays help, including generic Qt options
.SH "REPORTING BUGS"
Report bugs to https://github.com/lxqt/lxqt-config/issues
.SH "SEE ALSO"
\fBLXQt\fR has been tailored for users who value simplicity, speed, and
an intuitive interface, also intended for less powerful machines. See also:
.\" any module must refers to session app, for more info on start it
.P
\fBlxqt-config(1)\fR  LXQt Configuration Center


CHANGELOG
lxqt-config-2.1.0 / 2024-11-05
===============================
 * Fixed a FTBFS with Clang in `lxqt-config-input`.
 * Show unsupported message for `lxqt-config-input` under wayland.
 * Show unsupported message for `lxqt-config-monitor` under wlroots-based compositors.
 * Fixed a memory leak in `lxqt-config-input`.
 * Added nullity check for monitor config.
 * Safer size saving for the dialog of `lxqt-config-appearance`.
 * Do not load monitor or input settings under Wayland.
 * More informative tooltips in lxqt-config dialog.
 * Standardized desktop files for settings.
 * Consistent capitalization for GenericName for desktop files.
 * Added tooltip colors to LXQt Appearance.
 * Added checkbox to use palette provided by LXQt theme.
 * Fixed an old problem in setting wallpapers.
 * Don't start `xsettingsd` outside X11.
 * Added an option for toolbar icon size.
 * Make sure Monitor Settings dialog is shown completely on Wayland.
 * Remember size of Monitor Settings dialog.
 * Start `xsettingsd` on X11 only when GTK themes should be set.
 * Scale monitor images on resizing monitor settings dialog.

lxqt-config-2.0.0 / 2024-04-17
===============================
 * Ported to Qt6.
 * Improved the UI of monitor settings.
 * Corrected tab orders by reordering blocks of UI files.

lxqt-config-1.4.0 / 2023-11-05
===============================
 * Use lxqt-menu-data.
 * Fixed setting of primary monitor after a change in libkscreen.
 * Put cursor settings in ~/.Xresources (fixes the problem of not being able to set the cursor or its size when Xresources from `/etc` is read).
 * Corrected some typos.
 * Improved the message about default monitor settings.
 * Don't return uninitialized struct members.

lxqt-config-1.3.0 / 2023-04-15
===============================
 * Fixed potential segfault when loading keyboard layout settings.
 * lxqt-config-monitor: More header file inclusion for libkscreen 5.26.90.

lxqt-config-1.2.0 / 2022-11-05
===============================
 * Enhancements to theme list in appearance config dialog. Themes are always sorted alphabetically, user themes are distinguished, and theme folders can be opened by double clicking or by using context menus.

lxqt-config-1.1.0 / 2022-04-15
===============================
 * Updated `README.md`.
 * Fixed translation in "Remove Palettes" dialog of LXQt Appearance Configuration.
 * Removed Qt4 dead conditional code.
 * Removed unneeded KF5 KwindowSystem dependency.
 * Added a separate page for GTK style settings to LXQt Appearance Configuration.
 * Remember size of lxqt-config-input dialog.
 * Fixed manual editing of combos in LXQt Locale Configuration.
 * Fixed `mimeapps.list` path with empty `XDG_CURRENT_DESKTOP`.

lxqt-config-1.0.0 / 2021-11-04
===============================
 * Bumped minimum required Qt version to 5.15.
 * Do not force openbox configs under LXQt.
 * Made "Other Settings" work in Configuration Center.
 * Save and load Qt palettes.

lxqt-config-0.17.1 / 2021-04-16
===============================
  * Fixed compilation against Qt 5.12.

lxqt-config-0.17.0 / 2021-04-15
===============================
  * Dropped obsolete methods and fixed compiler warnings.
  * Completely moved to Qt5 signal/slot syntax.
  * Made lxqt-config-appearance work under Wayland.

lxqt-config-0.16.1 / 2020-11-14
===============================
  * Fixed a regression in setting the acceleration speed of mouse/touchpad.

lxqt-config-0.16.0 / 2020-11-01
===============================
  * Better UI for lxqt-config-brightness.
  * Avoid C++11 range for possible container detachment.
  * Only use QStringLiteral where it avoids allocations.
  * Initialize monitor settings config in "lxqt-config-monitor/monitor.h".
  * Use QFlags default constructor.
  * Fixed a crash in lxqt-config-monitor.
  * Fixed type associations and added "Default Application".
  * Detailed palette customization in lxqt-config-appearance.
  * Remember sizes of appearance and file associations dialogs.

lxqt-config-0.15.0 / 2020-04-22
===============================
  * Fixed changing of shortcut for switching keyboard layout.
  * Inserted default GTK3 themes in lxqt-config-appearance.
  * Set cursor style in GTK2, GTK3 and xsettingsd.
  * Don't use automatic string conversions.
  * C++11 code updates.
  * A better method to drag and drop monitors in lxqt-config-monitor.
  * Removed old mouse accel/sensitivity options.
  * Fixed warnings on applying settings in lxqt-config-monitor.
  * Fixed forgetting of lxqt-config-monitor's config.
  * Removed old mouse accel/sensitivity options.

lxqt-config-0.14.1 / 2019-02-24
===============================

  * Apply button for appearance and input config dialogs
  * Updated translations
  * Note for packagers: lxqt-config now depends on liblxqt >= 0.14.1

lxqt-config-0.14.0 / 2019-01-25
===============================

  * Implemented Features:
    - Introduced GTK appearance settings
    - Initial touchpad settings and improved input device handling
  * Implementation fixed:
    - lxqt-config: Removed ancient Qt workaround that creates problems now
    - lxqt-config-appearance: Some GUI fixes
    - lxqt-config-appearance: mkpath if settings of GTK doesn't exists.
    - lxqt-config-brightness: Backlight control updated
    - lxqt-config-brightness:
      XRandrBrightness::backlight_get_with_range return value
    - lxqt-config-monitor: Monitor name is added to output.
  * Improved cmake scripting
    - Set cmake_minimum_required to 3.1.0
    - Check for xi and libudev
    - Removed QUIET from find_package
    - Removed dependency on QtConcurrent
    - Removed locale compile definitions
    - Installed former missed man pages
  * Moved translations from lxqt-l10n back to lxqt-policykit
    - Removed obsolete translation fuctionality
    - Added translation promo in README.md
  * Translation updates

lxqt-config-0.13.0 / 2018-05-21
===============================

  * Spanish translation update
  * Update name for shortcut component

lxqt-config-input/lxqt-config-input.cpp
/*
    Copyright (C) 2013-2014  Hong Jen Yee (PCMan) <pcman.tw@gmail.com>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

#include <LXQt/SingleApplication>
#include <LXQt/ConfigDialog>
#include <LXQt/ConfigDialogCmdLineOptions>
#include <LXQt/Settings>
#include <QCommandLineParser>
#include <QMessageBox>
#include "mouseconfig.h"
#include "keyboardconfig.h"
#include "../liblxqt-config-cursor/selectwnd.h"
#include "keyboardlayoutconfig.h"

#ifdef WITH_TOUCHPAD
#include "touchpadconfig.h"
#include "touchpaddevice.h"
#endif

int main(int argc, char** argv) {
    LXQt::SingleApplication app(argc, argv);

    QString waylandMsg = QObject::tr("LXQt input settings are currently unsupported under Wayland.\n\nMouse, touchpad and keyboard can be configured in the settings of the compositor.");

    QCommandLineParser parser;
    LXQt::ConfigDialogCmdLineOptions dlgOptions;
    parser.setApplicationDescription(QStringLiteral("LXQt Config Input"));
    const QString VERINFO = QStringLiteral(LXQT_CONFIG_VERSION
                                           "\n\nliblxqt:   " LXQT_VERSION
                                           "\nQt:        " QT_VERSION_STR);
    app.setApplicationVersion(VERINFO);

    dlgOptions.setCommandLine(&parser);
#ifdef WITH_TOUCHPAD
    QCommandLineOption loadOption(QStringLiteral("load-touchpad"),
            app.tr("Load last touchpad settings."));
    parser.addOption(loadOption);
#endif
    parser.addVersionOption();
    parser.addHelpOption();
    parser.process(app);
    dlgOptions.process(parser);

    QByteArray configName = qgetenv("LXQT_SESSION_CONFIG");
    if(configName.isEmpty())
      configName = "session";
    LXQt::Settings settings(QString::fromUtf8(configName));

#ifdef WITH_TOUCHPAD
    bool loadLastTouchpadSettings = parser.isSet(loadOption);
    if (loadLastTouchpadSettings) {
        if (QGuiApplication::platformName() == QLatin1String("wayland"))
            qDebug().noquote() << waylandMsg; // Under wayland, show a debug message and exit.
        else
            TouchpadDevice::loadSettings(&settings);
        return 0;
    }
#endif

    if (QGuiApplication::platformName() == QLatin1String("wayland"))
    { // Show a message dialog under wayland if no options are given.
        QMessageBox::warning(nullptr, QObject::tr("Platform Unsupported"), waylandMsg);
        return 0;
    }

    LXQt::ConfigDialog dlg(QObject::tr("Keyboard and Mouse Settings"), &settings);
    dlg.setButtons(QDialogButtonBox::Apply|QDialogButtonBox::Close|QDialogButtonBox::Reset);
    dlg.enableButton(QDialogButtonBox::Apply, false); // disable Apply button in the beginning
    app.setActivationWindow(&dlg);

    LXQt::Settings qtSettings(QStringLiteral("lxqt"));

    /*** Mouse Config ***/
    MouseConfig* mouseConfig = new MouseConfig(&settings, &qtSettings, &dlg);
    dlg.addPage(mouseConfig, QObject::tr("Mouse"), QStringLiteral("input-mouse"));
    QObject::connect(&dlg, &LXQt::ConfigDialog::reset, mouseConfig, &MouseConfig::reset);
    QObject::connect(mouseConfig, &MouseConfig::settingsChanged, &dlg, [&dlg] {
        dlg.enableButton(QDialogButtonBox::Apply, true); // enable Apply button when something is changed
    });

    /*** Cursor Theme ***/
    SelectWnd* cursorConfig = new SelectWnd(&settings, &dlg);
    cursorConfig->setCurrent();
    dlg.addPage(cursorConfig, QObject::tr("Cursor"), QStringLiteral("preferences-desktop-theme"));
    QObject::connect(cursorConfig, &SelectWnd::settingsChanged, &dlg, [&dlg] {
        dlg.enableButton(QDialogButtonBox::Apply, true);
    });

    /*** Keyboard Config ***/
    KeyboardConfig* keyboardConfig = new KeyboardConfig(&settings, &qtSettings, &dlg);
    dlg.addPage(keyboardConfig, QObject::tr("Keyboard"), QStringLiteral("input-keyboard"));
    QObject::connect(&dlg, &LXQt::ConfigDialog::reset, keyboardConfig, &KeyboardConfig::reset);
    QObject::connect(keyboardConfig, &KeyboardConfig::settingsChanged, &dlg, [&dlg] {
        dlg.enableButton(QDialogButtonBox::Apply, true);
    });

    /*** Keyboard Layout ***/
    KeyboardLayoutConfig* keyboardLayoutConfig = new KeyboardLayoutConfig(&settings, &dlg);
    dlg.addPage(keyboardLayoutConfig, QObject::tr("Keyboard Layout"), QStringLiteral("input-keyboard"));
    QObject::connect(&dlg, &LXQt::ConfigDialog::reset, keyboardLayoutConfig, &KeyboardLayoutConfig::reset);
    QObject::connect(keyboardLayoutConfig, &KeyboardLayoutConfig::settingsChanged, &dlg, [&dlg] {
        dlg.enableButton(QDialogButtonBox::Apply, true);
    });

#ifdef WITH_TOUCHPAD
    /*** Touchpad Config ***/
    TouchpadConfig* touchpadConfig = new TouchpadConfig(&settings, &dlg);
    dlg.addPage(touchpadConfig, QObject::tr("Mouse and Touchpad"), QStringLiteral("input-tablet"));
    QObject::connect(&dlg, &LXQt::ConfigDialog::reset, touchpadConfig, &TouchpadConfig::reset);
    QObject::connect(touchpadConfig, &TouchpadConfig::settingsChanged, &dlg, [&dlg] {
        dlg.enableButton(QDialogButtonBox::Apply, true);
    });
#endif

    // apply all changes on clicking Apply
    QObject::connect(&dlg, &LXQt::ConfigDialog::clicked, [=, &dlg] (QDialogButtonBox::StandardButton btn) {
        if (btn == QDialogButtonBox::Apply)
        {
            mouseConfig->applyConfig();
            cursorConfig->applyCusorTheme();
            keyboardConfig->applyConfig();
            keyboardLayoutConfig->applyConfig();
#ifdef WITH_TOUCHPAD
            touchpadConfig->applyConfig();
#endif
            // disable Apply button after changes are applied
            dlg.enableButton(btn, false);
        }
        else if (btn == QDialogButtonBox::Reset)
            dlg.enableButton(QDialogButtonBox::Apply, false); // disable Apply button on resetting too
    });

    dlg.setWindowIcon(QIcon::fromTheme(QStringLiteral("input-keyboard")));

    LXQt::Settings mConfigInputSettings(QStringLiteral("lxqt-config-input"));
    dlg.resize(mConfigInputSettings.value(QStringLiteral("size")).toSize().expandedTo(QSize(600, 400)));

    const QString initialPage = dlgOptions.page();
    if (!initialPage.isEmpty())
        dlg.showPage(initialPage);

    dlg.exec();
    mConfigInputSettings.setValue(QStringLiteral("size"), dlg.size());
    return 0;
}
